modules/filetype.nix

```
@@ -52,7 +54,9 @@ in
         pattern = filetypeDefinition "set filetypes matching the specified pattern";
       };
 
-  config.extraConfigLua = helpers.mkIfNonNull' config.filetype ''
-    vim.filetype.add(${helpers.toLuaObject config.filetype})
-  '';
+  config.extraConfigLua =
+    lib.mkIf (cfg != null && (cfg.extension != null || cfg.filename != null || cfg.pattern != null))
```

This is ugly, but it is probably the right thing to do.

---

Would it make sense to do something like `all (v: v == null) (attrValues config fileType)`?

I think even if users had freeform definitions, `toLuaObject` would still ignore nulls.

Might need to also check `config.filetype ? __empty` though, in case the user explicitly wants to define an empty filetype?

(Is the option even freeform? If not some of these caveats don't apply. Sorry for the poor review, I'm on my phone atm!)

---

I think it warrants a library function similar to `mkIfNonNull` and `mkIfNonNull'` that test attrs similar to `toLuaObject`. If `toLuaObject` would produce something empty (it can be configured via function params what is considered empty: only if nil, or if empty table, or if empty string), condition is false. I don't think I know all this options and combinations with __empty, __raw etc to write one though.

---

```suggestion
    lib.mkIf (cfg != null && (lib.any (v: v != null) (lib.attrValues cfg)))
```

I've had a closer look, and it should be fine here to check whether any attrs are non-null using the builtins [`any`](https://noogle.dev/f/builtins/any) and [`attrValues`](https://noogle.dev/f/builtins/attrValues).

Since this isn't a freeform option we don't have to worry about unexpected attrs or `__empty`.

---

Thanks, that's nicer.